File: Godot/XRPlayer.cs
namespace Game5;

using Godot;

public partial class XrPlayer : Node3D, IMaterializeable, IPlayer
{
	private const float FallAcceleration = 9.8f;
	public XrPlayerController Controller;
	private Mesh _mesh;
	private bool _materialized = true;

	/*
	public override void _Ready()
	{
		base._Ready();
		Dematerialize();
	}
	*/

	public override void _Process(double delta)
	{
		base._Process(delta);
		Controller.ProcessControls();
	}

	public override void _PhysicsProcess(double delta)
	{
		base._PhysicsProcess(delta);
		var playerBody = GetNode<CharacterBody3D>("XROrigin3D/PlayerBody");
		var velocity = playerBody.Velocity;
		
		if (_materialized && !playerBody.IsOnFloor())
		{
			velocity.Y -= FallAcceleration * (float)delta;
		}

		playerBody.Velocity = velocity;
		
		playerBody.MoveAndSlide();
	}

	public void Materialize()
	{
		_materialized = true;
		GetNode<CollisionShape3D>("XROrigin3D/PlayerBody/CollisionShape3D").Disabled = false;
		GetNode<MeshInstance3D>("XROrigin3D/PlayerBody/Pivot/MeshInstance3D").Mesh = _mesh;
	}
	
	public void Dematerialize()
	{
		_materialized = false;
		GetNode<CollisionShape3D>("XROrigin3D/PlayerBody/CollisionShape3D").Disabled = true;
		_mesh = GetNode<MeshInstance3D>("XROrigin3D/PlayerBody/Pivot/MeshInstance3D").Mesh;
		GetNode<MeshInstance3D>("XROrigin3D/PlayerBody/Pivot/MeshInstance3D").Mesh = null;
	}

	public bool IsMaterialized()
	{
		return _materialized;
	}
}

File: Sources/Game/XrPlayer.swift
//
//  XrPlayer.swift
//
//
//  Created by Anthony Michalek on 2/18/24.
//

import SwiftGodot;

@Godot
class XrPlayer: Node3D, PlayerProtocol {
    private final var fallAcceleration: Float = 9.8;
    private var controller: XrPlayerControllerProtocol?;
    private var mesh: Mesh?;
    private var playerBody: CharacterBody3D?;
    private var materialized: Bool = true;
    
    override func _ready() {
        super._ready();
        // dematerialize();
        controller = XrPlayerController(xrPlayer: self);
        playerBody = getNode(path: "XROrigin3D/PlayerBody") as! CharacterBody3D;
    }
    
    override func _process(delta: Double) {
        super._process(delta: delta);
        controller?.processControls();
    }
    
    override func _physicsProcess(delta: Double) {
        if Engine.isEditorHint() { return }
        if (materialized && playerBody!.isOnFloor())
        {
            playerBody!.velocity.y -= fallAcceleration * Float(delta);
        }
        
        playerBody!.moveAndSlide();
        super._physicsProcess(delta: delta);
        
    }
    
    func materialize() {
        materialized = true;
        (playerBody!.getNode(path: "CollisionShape3D") as! CollisionShape3D).disabled = false;
        (playerBody!.getNode(path: "Pivot/MeshInstance3D") as! MeshInstance3D).mesh = mesh;
    }
    
    func dematerialize() {
        materialized = false;
        (playerBody!.getNode(path: "CollisionShape3D") as! CollisionShape3D).disabled = true;
        mesh = (playerBody!.getNode(path: "PlayerBody/Pivot/MeshInstance3D") as! MeshInstance3D).mesh;
        (playerBody!.getNode(path: "Pivot/MeshInstance3D") as! MeshInstance3D).mesh = nil;
    }
    
    func isMaterialized() -> Bool {
        return materialized;
    }
    
    func getController() -> PlayerControllerProtocol {
        return controller!;
    }
    
}
